### ISSUE TITLE
idea: More memory efficient watch cache

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#90179]
The current watch cache requires apiserver hold at least every item in a collection in memory (otherwise it cannot serve lists from the cache), and more (a history window). This is a large amount of memory consumed.

There's an alternative design that IIRC wasn't possible with etcd2 but should now be possible: catch-up watches.

Instead of serving a list from the cache,
* we'd serve it directly from etcd if it is paginated (which we already do), since paginated requests are relatively cheap.
* But unpaginated list requests will have a max frequency, e.g. 1 / second. All requests for the same list get held for either that amount of time, or until an in-flight etcd list query finishes. Then we serve the same data to all requesting clients. (To make this memory efficient, we may have to do a series of paginated requests against etcd, and stream the results to clients.) This greatly penalizes unpaginated lists, and that's OK IMO.

We'd retain the part of the watch cache that multiplexes a single watch to many watchers. We keep all these watchers at exactly the same revision; as a change comes in, we broadcast it to all watchers.

Now, for a new watch request, we have a problem-- the request might be for revision X, when all the existing watchers are on revision Y.
* If X > Y, add the watcher, and just skip events until the existing watchers catch up.
* If X < Y, add the watcher and begin queueing (not sending) events; do a query from etcd between revisions X and Y (e.g., this could be a watch that we terminate early!). If that's successful, send those to the watcher, then the queued events (if any), then the watcher is caught up.

This is inefficient if there's many revisions that need to be processed for every new watcher. We can help this by keeping a small cache of revisions, so that we have to do fewer and smaller revision queries from etcd (this cache can be arbitrarily small, unlike the cache for supporting LISTs). Alternatively, we could batch these as suggested for lists above.

Here's some challenges with this:
* boundary conditions: lots of opportunity to have an off-by-one error when joining watchers or batching catch-up watch requests. Exhaustive testing can fix this.
* filters: lists from etcd and catch-up watch requests could well end up throwing away much of the events for not matching a filter. This is inefficient and would be worse than the existing design for these queries (the existing design is already pretty bad--aside from targeted queries which have been optimized). The only real (IMO) solution for this problem is to push the selection behavior into the storage layer, but that's super hard.

TL;DR: Our existing cache is about locally reproducing the watch behavior. This proposal is about changing the theoretical nature of this, so that instead we think of it in terms of joining up requests that can share work. It's no longer primarily a cache, it's a smart [singleflight](https://godoc.org/golang.org/x/sync/singleflight).

There's a memory/cpu tradeoff here, at least as long as apiserver has to do the filtering. So, I'm not claiming we should do this immediately, I just want to document the approach. This would let apiserver scale to bigger cluster sizes (where not everything fits in RAM). Which is another way of saying it'd permit some amount of horizontal scaling (add more machines) instead of vertical scaling (add more RAM).

/sig api-machinery
/priority awaiting-more-evidence

**Comment by lavalamp** [comment]
A simpler, half measure for demonstrating whether or not this has potential would be to just stop serving lists out of cache, but keep the watch history window and leave the watch serving mechanism the same. For big collections, the history window will be vastly smaller than the number of objects in the collection.

**Comment by wojtek-t** [comment]
> we'd serve it directly from etcd if it is paginated (which we already do), since paginated requests are relatively cheap.

This is the perception that I'm trying to change, but so far spectacularly fail on that :)
So to clarify - yes, listing a single page from etcd is not extremely expensive. But imagine a 5k-node cluster with 150k pods (which is what we support) and a Kubelet that was restarted and has to initialize. It has to list all pods from its own node. So what it would have to do if we wouldn't have watch cache is:
- list (in pages, but still) all pods form the etcd, deserialize them, filter out 99,98% of pods that it's not interested in, drop them and send to the watcher the remaining 0.02%
Those deserializations (and memory allocations) that this would require, as well as some network bandwidth between etcd and apiserver (there are many environments that run etcd on a separate set of machines) is not something we can easily forget.

So I guess what I'm saying, if most of lists would be interested in majority of contents if we eliminate selectors, that would be fine. But that's not the case.


> This is inefficient and would be worse than the existing design for these queries (the existing design is already pretty bad--aside from targeted queries which have been optimized). 

I don't agree it's bad. It's usecase-oriented. We don't optimize things that aren't frequently used - i.e. their consequences aren't that bad.

> The only real (IMO) solution for this problem is to push the selection behavior into the storage layer, but that's super hard.

Agree (both that it's the solution and that it's hard).

> This would let apiserver scale to bigger cluster sizes (where not everything fits in RAM). Which is another way of saying it'd permit some amount of horizontal scaling (add more machines) instead of vertical scaling (add more RAM).

I agree. But there are also other things that we can do - i.e. not serve all the resources from the same set of apiservers. e.g. we can serve resource X from apiservers 1, 2 and 3, and other resources from apiservers 4, 5 and 6. Sure - that doesn't solve the problem where there is huge number of resources of a single type.
But I don't think memory is really that big problem - at least for GKE is not a problem at all.


> A simpler, half measure for demonstrating whether or not this has potential would be to just stop serving lists out of cache, but keep the watch history window and leave the watch serving mechanism the same. For big collections, the history window will be vastly smaller than the number of objects in the collection.

We accidentally kind-of did that in https://github.com/kubernetes/kubernetes/issues/86483. That we couldn't even upgrade any large cluster with that. We are very far from being able to do that.

**Comment by smarterclayton** [comment]
I'll note here I was considering something that is maybe the logical inverse of this:

1. replace both live etcd read and watch cache with a single watch fed cache on all etcd objects that are part of the kube model
2. perform decode from storage exactly one time per object revision
3. allow logical indices on that cache (cluster, resource, multiple resources, namespaces) to support other types of queries
4. if you need to do consistent reads or scans at a version, do key only reads from etcd and fill them (wait if necessary) from your cache

That reduces decode churn (a significant portion of memory usage) and also sets up improvements to uniform watching across types, etc.  I think that would make us more likely to fit in less memory although it does not get to letting us bypass memory.

**Comment by logicalhan** [comment]
> I'll note here I was considering something that is maybe the logical inverse of this:
> 
> 1. replace both live etcd read and watch cache with a single watch fed cache on all etcd objects that are part of the kube model
> 2. perform decode from storage exactly one time per object revision
> 3. allow logical indices on that cache (cluster, resource, multiple resources, namespaces) to support other types of queries
> 4. if you need to do consistent reads or scans at a version, do key only reads from etcd and fill them (wait if necessary) from your cache
> 
> That reduces decode churn (a significant portion of memory usage) and also sets up improvements to uniform watching across types, etc. I think that would make us more likely to fit in less memory although it does not get to letting us bypass memory.

I'm not crazy about this approach. From my read, it sounds like you basically want an etcd-lite living inside apiserver, bringing along with it other things we'd have to be worried about, like compacting MVCC versions in the apiserver cache. Given the possibility that a user can have multiple etcds configured for different resource types, the complexity of this approach makes it considerably less attractive to me than @lavalamp's idea. And I'm not saying this because I work with @lavalamp, it is just how I genuinely feel when weighing these two ideas side-by-side. @lavalamp's approach appeals to me because it decreases internal state in the apiserver (making it easier to reason about, for me anyway), decreases apiserver's memory footprint and leaves all the MVCC logic to be handled by the layer which specializes in handling it (i.e. etcd).

**Comment by lavalamp** [comment]
> replace both live etcd read and watch cache with a single watch fed cache on all etcd objects that are part of the kube model

How can this work when collections may be in distinct etcd clusters? (events commonly configured that way)

> allow logical indices on that cache (cluster, resource, multiple resources, namespaces) to support other types of queries

I don't want to write a database, do you? (actually that sounds really fun, I just don't think this is the place)

I think an etcd / watchcache layer that understood the object format and could do the right conversions (so that apiserver can pass through bytes rather than decoding watch events) would be interesting. watches and other api requests have different scaling characteristics and it'd be useful to be able to provision them differently. I don't really want to make the watch cache more feature-ful unless we can isolate it more thoroughly from the rest of the code.

**Comment by smarterclayton** [comment]
> How can this work when collections may be in distinct etcd clusters? (events commonly configured that way)

They can be configured that way but I'm not sure that's where the value is.  Sharding by namespace is out.  Sharding by resource is.... not that great.  In production across the clusters i have data on at least (post the event spam fixes) events are not even a plurality of the write rate of a cluster.

> I don't want to write a database, do you? (actually that sounds really fun, I just don't think this is the place)

Watch cache is half a database.  Our informers are half a database.  I think tipping them over to be just a slight bit closer to a database would let people build the kinds of integrations they need to do.  The average component operator I see has over a dozen resource watches.  I'd like to be able to watch all the things happening in a namespace for user focused consoles.  Etc.

>  I don't really want to make the watch cache more feature-ful unless we can isolate it more thoroughly from the rest of the code.

When wojtek and I discussed it I was describing it as "rip out all infrastructure that supports get, list, and watch and unify it into one set of code with a very clean fallback path if necessary".  It wouldn't be a layer on top of storage, it would replace the storage impl for those methods.  I don't think watch scaling is the problem anymore - it's multi-gigabyte list reads in large clusters, and to solve that you have to go after the expensive bits.  I think list and watch, unified with watch cache and a clear ability to leverage the cache for lists, replacing the storage call through, is potentially the way to do that.  Again, this is mostly thought experiment, but I want to reduce complexity by unifying components and getting clear wins.

Instead of thinking about watch cache and storage as separate, imagine we had a subcomponent cache that could be shared between informers and storage equally and efficiently across multiple types of objects that can be fed from client apis or an etcd storage object.

Other areas where this intersects in the long run is kubelet surviving reboot while partitioned.  Every integration on the node has broken the ability of kubelet to do any sort of "hack remembering pods", and I'd argue we shouldn't try, but instead let a downstream component keep a consistent view of the upstream.  Ideally the same code we'd use to build those consistent views could be replicated in many places.

**Comment by lavalamp** [comment]
Are you trying to build memcached?

At least part of what you're saying overlaps with what I want, I think.

> I don't think watch scaling is the problem anymore - it's multi-gigabyte list reads in large clusters, and to solve that you have to go after the expensive bits.

I think there's much easier solutions, like techniques that permit lists to be partially successful and latch in some progress, so the client can just keep trying and eventually successfully get everything. See this issue's sibling: https://github.com/kubernetes/kubernetes/issues/90339

I'm happy to sync up in realtime to see if there's something in this space that meets the right benefit : effort ratio. Right now I don't think there is?

**Comment by smarterclayton** [comment]
> Are you trying to build memcached?

Nope.  Kube is a distributed system.  A core pattern we have is a consistent replicated state cache (not quite a viewstamped replicated state machine).  Controllers are trying to select sets and subsets of full cluster content efficiently, then offer the same efficiency to local consumers.  Kubelets, the namespace controller, and the gc controller are "mega watchers" that have to watch everything-ish, or a subset of everything ish.  Rather than model apiservers as one type of cache and controllers with a different sort of cache, look for underlying list/watch patterns that scale to both of them.

I don't have an issue handy for it right now but I have seen large clusters already push past the point where a single resource list won't fit in memory and a paged query takes too long to fit within compaction.  Instead of coming up with different semantics for existing calls, I was thinking more along the lines of how we could collapse some of the built up abstractions.  I'm not arguing that to address 90339 you have to fix it this way - I'm just documenting a potentially different direction in general - keep the multi-watcher logic, but cut out intervening layers and unify, then look for ways to align that investment with controller caches and better object selection.

**Comment by lavalamp** [comment]
> Rather than model apiservers as one type of cache and controllers with a different sort of cache, look for underlying list/watch patterns that scale to both of them.

They do pretty different things, so I'm a little skeptical that there's a big win to be found here. If we were going to make a write-through cache of some sort then this would make more sense to me.

> I have seen large clusters already push past the point where a single resource list won't fit in memory and a paged query takes too long to fit within compaction

Yeah the cluster can fall off a cliff here. I think we should explore tactical fixes before architectural ones. You don't even have to get as fancy as #90339, honestly we could fix it client side with some clever use of the continuation tokens.

**Comment by lavalamp** [comment]
> You don't even have to get as fancy as #90339, honestly we could fix it client side with some clever use of the continuation tokens.

I wrote this up since it's probably not obvious: https://github.com/kubernetes/kubernetes/issues/98541

**Comment by lavalamp** [comment]
Yeah, that was my point, I don't think we do need a write-through cache (or anything in the category), and therefore I don't see the benefit of a big architectural change :)

**Comment by smarterclayton** [comment]
What I was describing is the watch cache but better integrated and more efficient.  The more efficient is the key point.

**Comment by wojtek-t** [comment]
@smarterclayton 

> I don't think watch scaling is the problem anymore - it's multi-gigabyte list reads in large clusters, and to solve that you have to go after the expensive bits like decode.

The former (watch scaling) is still a problem too. We just have this new problem of large lists.

So I actually think that the original part of watchcache (the serving watches part itself) is still needed. But I agree that it should be possible to put it very clearly on top of the cache you're proposing and it should also simplify this part (though I don't think we will be able to completely remove it).


> Why do we need a write through cache?

uff - that's something I would be really against


> What I was describing is the watch cache but better integrated and more efficient. The more efficient is the key point.

I generally agree. When I started with watchcache 5 years ago, it was supposed to only serve watches. And then we added lists. And  it was evolving and evolving and we're where we are. Knowing everything we know now, we would do that differently now.

I don't perceive this as a "huge architectural change". I rather thing about it as rewriting a single layer of apiserver. We shouldn't touch storage interface, we don't have to touch the etcd client. So we only have to rewrite the in-between stuff. But the API is well defined on both ways of that layer.
My only concern here is connected to introducing new bugs; we had a number of them in this layer in the past and if we decide to effectively rewrite this layer from scratch, we can introduce them (or other) again. We have a bunch of tests, that many of them are unit tests, which will have to change, so the risk is non-negligible imho.

**Comment by smarterclayton** [comment]
Yeah, I think that's a key point.  We don't have a ton of tests that verify correctness at the storage level (conformance of specific behavior) and that complicates a bit of this, but an effort to grow those tests that LED to us then being able to go after some bigger wins / try out more options with confidence would be my desired state.  That can be separated from the short term tactical things we need to do - not trying to block those.

The larger thing would be about what semantics of etcd we want to preserve, where do we want to go, do we want to allow broader use of the kube-apiserver core divorced from kube (i.e. i'm trying to explore some alternate ways of doing tenancy with clusters that would benefit from reassessing the storage layer), do we think we might need to improve watches of many resources by adding new apis, should we improve filtering and field selection, etc.  Just spitballing though.

**Comment by Jeffwan** [comment]
>  Sharding by namespace is out.

@smarterclayton   Could I know why this is not viable? Trying to understand the bottlenecks to go to bigger cluster size like 50k.

**Comment by dims** [comment]
cc @nikhita @palnabarun

**Comment by julienlau** [comment]
Naïve suggestion following this very interesting discussion : why everything should be stored in etcd ?
Maybe the control plane storage backend should be extended with another engine SQL or NewSQL (or maybe an event broker) to segregate some of etcd workload ? This may also enable to crank down some lease timeout and have large cluster more reactive ? 
 

Back in the days DC/OS used CockroachDB to store secrets and not the zookeeper cluster that managed DC/OS cluster state.
I also observe that Rancher is exploring swapping etcd by an RDBMS for K3S control plane (Kine https://rancher.com/blog/2020/scaling-fleet-kubernetes-million-clusters )

**Comment by serathius** [comment]
I'm currently drafting a design proposal that bridges the ideas discussed here. By explicitly decoupling the watch demux from the storage layer, we can move towards @lavalamp's vision of a 'smart singleflight' multiplexer, while fully maintaining the in-memory LIST semantics required for performance.

This separation reduces lock contention between readers and writes. I've already successfully implemented this architecture over in the etcd project cache library, including robustness testing for correctness under strong consistency requirements.

I'll be sharing the full design document soon, stay tuned!
